### ISSUE TITLE
Incremental compiler misses change to value class, and results to NoSuchMethodError at runtime

### ISSUE THREAD

**Issue by xuwei-k** [sbt/sbt#1171]
see https://github.com/xuwei-k/sbt-value-class-bug
### step

From sbt shell, `compile` A.scala (without value class)

``` scala
case class A(a: String)
```

B.scala:

``` scala
object B {
  def foo = A("a")
}
```

Change to A.scala to following:

``` scala
case class A(a: String) extends AnyVal
```

From sbt shell, `run` (without `clean` )
### problem

```
[info] [error] (run-main-0) java.lang.NoSuchMethodError: B$.foo()LA;
[info] java.lang.NoSuchMethodError: B$.foo()LA;
[info]  at C$delayedInit$body.apply(C.scala:3)
[info]  at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
[info]  at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
[info]  at scala.App$$anonfun$main$1.apply(App.scala:71)
[info]  at scala.App$$anonfun$main$1.apply(App.scala:71)
[info]  at scala.collection.immutable.List.foreach(List.scala:318)
[info]  at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
[info]  at scala.App$class.main(App.scala:71)
[info]  at C$.main(C.scala:1)
[info]  at C.main(C.scala)
[info]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
[info]  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]  at java.lang.reflect.Method.invoke(Method.java:606)
[info] [trace] Stack trace suppressed: run last compile:run for the full output.
[info] java.lang.RuntimeException: Nonzero exit code: 1
[info]  at scala.sys.package$.error(package.scala:27)
[info] [trace] Stack trace suppressed: run last compile:run for the full output.
[info] [error] (compile:run) Nonzero exit code: 1
[info] [error] Total time: 4 s, completed Mar 11, 2014 9:32:49 AM
[error] x value-class-bug / simple
```
### notes

`NoSuchMethodError` does not occur if `clean` is executed first.

**Comment by eed3si9n** [comment]
Do you know which sbt or Scala version did this behavior start?

**Comment by xuwei-k** [comment]
> Do you know which sbt or Scala version did this behavior start?

Maybe from the beginning. following all combinations failed.

https://github.com/xuwei-k/sbt-value-class-bug/tree/3e9bb5b83
https://travis-ci.org/xuwei-k/sbt-value-class-bug/builds/20529146

sbt
0.11.3, 0.12.0, 0.12.1, 0.12.2, 0.12.3, 0.12.4, 0.13.0, 0.13.1, 0.13.2-M2

scala
2.10.0, 2.10.1, 2.10.2, 2.10.3, 2.10.4-RC3, 2.11.0-RC1

**Comment by jsuereth** [comment]
@gkossakowski This looks like something maybe for 0.13.3.  I'll let you figure out when you think we should tackle it.

**Comment by gkossakowski** [comment]
Not sure I understand. This is a bug in Scala compiler and not sbt, right?

**Comment by gkossakowski** [comment]
Oh sorry, I misread it. It's a genuine bug.

**Comment by gkossakowski** [comment]
I looked into this and the problem is quite fundamental. Incremental compiler detects changes to APIs (and triggers recompilation) based on class/method signatures as viewed at pickling phase. Pickling phase represents signatures before erasure. That means, if you have:

``` scala
class A
def foo: A = ???
```

and you change `class A` to be a value class the signature of method `foo` won't change. It will be still `foo: A`. However, the erasure of `foo` is very different if `A` is a value class (it erases to whatever value class wraps). Given the fact that incremental compiler looks only at pre-erasure types, it has no why of noticing the difference.

The most principled solution would be to index all signatures twice:
- at pickling phase
- after erasure (or posterasure, so value classes are completely erased)

We would have to introduce a whole new tree walk that indexes all the types. However, that would be very expensive to do especially given the fact that #1078 is not fixed yet.

The alternative would be to introduce a special case for value classes and index references to them separately. Specifically, when indexing TypeRefs, we would capture the fact whether a class is a value class. This solution is a bit ugly but cheap to implement and execute (performance overhead would be minimal).

**Comment by gkossakowski** [comment]
This is definitively 0.13.3 material.

**Comment by gkossakowski** [comment]
Scheduled for sbt 0.13.6.

**Comment by retronym** [comment]
Rather than introducing a new compiler phase to capture signatures at erasure, it would be possible to modify the existing API collection phase to look compute the erased type signature of each method it looks at.

```
scala> :power
** Power User mode enabled - BEEP WHIR GYVE **
** :phase has been set to 'typer'.          **
** scala.tools.nsc._ has been imported      **
** global._, definitions._ also imported    **
** Try  :help, :vals, power.<tab>           **

scala> class VC(val self: String) extends AnyVal; class C { def foo(vc: VC) }
<console>:53: error: class C needs to be abstract, since method foo is not defined
       class VC(val self: String) extends AnyVal; class C { def foo(vc: VC) }
                                                        ^

scala> class VC(val self: String) extends AnyVal; class C { def foo(vc: VC) = 0 }
defined class VC
defined class C

scala> val foo = typeOf[C].member(TermName("foo"))
foo: $r.intp.global.Symbol = method foo

scala> exitingTyper(foo.info) // this is the type signature currently used in the SBT's index.
res3: $r.intp.global.Type = (vc: VC)Int

scala> exitingPostErasure(foo.info) // we'd need to grab this one, too
res5: $r.intp.global.Type = (vc: String)Int
```

This approach is much less expensive than a new compiler pass, as you are just doing work more eagerly that would be done later on in the compiler anyway. There is a risk that changing the order that this happens could trigger a data race in the compiler, but some testing with a few big projects would likely flush out that sort of problem.

**Comment by xuwei-k** [comment]
Should I close this?

**Comment by eed3si9n** [comment]
> Should I close this?

We should probably wait till the fix gets forward ported into sbt/incrementalcompiler.

**Comment by dwijnand** [comment]
So this is fixed in 0.13.9-RC1. I saw that I had closed this by pushing latest 0.13 to the sbt-zero-thirteen repo, so I re-opened it. But I'm going to re-close it as we have pull requests for making sure this is forward-porting to 1.0.
